test/jdk/java/awt/TrayIcon/DisposeInActionEventTest/DisposeInActionEventTest.java

```diff
@@ -21,76 +21,69 @@
-                    "  don't see a tray icon, please, make sure that the tray area",
-                    "  (also called Taskbar Status Area on MS Windows, Notification",
-                    "  Area on Gnome or System Tray on KDE) is visible.",
-                    "Double-click with " + clickInstruction + " button on the tray icon to trigger the",
-                    "  action event. Brief information about action events is printed",
-                    "  below. After each action event the tray icon is removed from",
-                    "  the tray and then added back in a second.",
-                    "The test performs some automatic checks when removing the icon. If",
-                    "  something is wrong the corresponding message is displayed below.",
-                    "  Repeat double-clicks several times. If no 'Test FAILED' messages",
-                    "  are printed, press PASS button else FAIL button."
-            };
+    public static void main(String[] args) throws Exception {
+        if (!SystemTray.isSupported()) {
+            throw new jtreg.SkippedException("The test cannot be run because " +
+                    "SystemTray is not supported.");
         }
-        Sysout.createDialogWithInstructions(instructions);
-    }
+        String clickInstruction = (Platform.isOSX()) ? "Right-click" : "Double-left click";
+
+        String instructions = "When the test starts, it adds the icon to the tray area. If you\n" +
+                       "  don't see a tray icon, please, make sure that the tray area\n" +
+                       "  (also called Taskbar Status Area on MS Windows, Notification\n" +
+                       "  Area on Gnome or System Tray on KDE) is visible.\n\n" +
+                        clickInstruction + " the button on the tray icon to trigger the\n" +
+                       "  action event. Brief information about action events is printed\n" +
+                       "  in the frame. After each action event the tray icon is removed from\n" +
+                       "  the tray and then added back in a second.\n\n" +
+                       "The test performs some automatic checks when removing the icon. If\n" +
+                       "  something is wrong the corresponding message is displayed below.\n" +
+                       "  Repeat clicks several times. If no 'Test FAILED' messages\n" +
+                       "  are printed, press PASS button else FAIL button.";
 
-    @Override
-    public void start() {
-        setSize(200, 200);
-        setVisible(true);
-        validate();
+        PassFailJFrame.builder()
+                .title("Test Instructions Frame")
+                .instructions(instructions)
+                .testTimeOut(10)
+                .rows(15)
+                .columns(45)
+                .testUI(DisposeInActionEventTest::showFrameAndIcon)
+                .build()
+                .awaitAndCheck();
```

Comment:
I propose removing the icon from the system tray in a `finally` block. Otherwise, the test doesn't exit if you run it without jtreg.

Reply:
i'm not sure i understand this, the test is trying to remove right after the click action is delivered and not before the test is closing, so wouldn't putting the removal of the icon in a finally block defeat the purpose of the test?

Reply:
@alisenchung  Since the trayicon is added back after removal and PassFailJFrame handles only disposal of test windows, Aleksei is suggesting we handle the removal of trayicon explicitly something like below. You probably want to move the `systemTray` and `trayIcon` vars accordingly.

```
try {
    PassFailJFrame.builder()
                  .title("Event Message Display")
                  .instructions(instructions)
                  .testTimeOut(5)
                  .rows(15)
                  .columns(45)
                  .testUI(DisposeInActionEventTest::showFrameAndIcon)
                  .build()
                  .awaitAndCheck();
} finally {
     systemTray.remove(trayIcon);
}

```

Reply:
That's correct! Otherwise, the orphan icon is left the system tray. When the test is run without jtreg, it doesn't terminate because the tray icon doesn't allow EDT to stop.

Reply:
Alternatively, you can remove the system tray icon in `windowClosing` or `windowClosed` event of `WindowListener` associated with the frame.